packages/react-dom-bindings/src/client/ReactDOMHostConfig.js

```diff
@@ -845,42 +859,210 @@ export function isHydratable(type: string, props: Props): boolean {
   }
 }
 
-export function canHydrateInstance(
+// In certain contexts, namely <body> and <head>, we want to skip past Nodes that are in theory
+// hydratable but do not match the current Fiber being hydrated. We track the hydratable node we
+// are currently attempting using this module global. If the hydration is unsuccessful Fiber will
+// call getLastAttemptedHydratable which uses this cursor to return the expected next
+// hydratable.
+let hydratableNode: null | HydratableInstance = null;
+
+export function getLastAttemptedHydratable(): null | HydratableInstance {
+  return hydratableNode;
+}
+
+export function getNextMatchingHydratableInstance(
   instance: HydratableInstance,
   type: string,
   props: Props,
+  rootOrSingletonContext: boolean,
```

Comment:
It would be better to split all these into two functions. This function can't do anything useful with it, other than maybe save code but it can do that with helper functions too.

Where as the parent can possibly create two branches of code that only needs to check this flag once or even not make it a flag at all sometimes where it's just implied by the code path.

Reply:
addressed this in refactor of hydrationcontext for #26256 should be good now